packages/react-router/lib/components.tsx

```diff
@@ -152,6 +163,307 @@ export function createMemoryRouter(
   }).initialize();
 }
 
+class Deferred<T> {
+  status: "pending" | "resolved" | "rejected" = "pending";
+  promise: Promise<T>;
+  // @ts-expect-error - no initializer
+  resolve: (value: T) => void;
+  // @ts-expect-error - no initializer
+  reject: (reason?: unknown) => void;
+  constructor() {
+    this.promise = new Promise((resolve, reject) => {
+      this.resolve = (value) => {
+        if (this.status === "pending") {
+          this.status = "resolved";
+          resolve(value);
+        }
+      };
+      this.reject = (reason) => {
+        if (this.status === "pending") {
+          this.status = "rejected";
+          reject(reason);
+        }
+      };
+    });
+  }
+}
+
+// Copied from react-dom types
+export interface RouterProviderProps {
+  router: RemixRouter;
+  flushSync?: (fn: () => unknown) => undefined;
+}
+
+/**
+ * Given a Remix Router instance, render the appropriate UI
+ */
+export function RouterProvider({
+  router,
+  flushSync: reactDomFlushSyncImpl,
```

Comment:
This is the same implementation as before, copied into here but with flushSync dependency injected via a prop